keps/sig-network/2433-topology-aware-hints/README.md

```diff
@@ -0,0 +1,578 @@
+
+## Design Details
+
+### Assumptions
+
+- Incoming traffic is proportional to the number of allocatable CPU cores in a
+  zone. Although this is an imperfect metric, it is the best available way of
+  predicting how much traffic will be received in a zone. If we are unable to
+  derive the number of allocatable cores in a zone we will fall back to the
+  number of nodes in that zone.
+- Service capacity is proportional to the number of endpoints in a zone. This
+  assumes that each endpoint has equivalent capacity. Although this is not
+  always true, it usually is. We can explore ways to deal with variable capacity
+  endpoints in the future.
+
+### Configuration
+
+A new `topology.kubernetes.io/topology-aware-routing` annotation will be
+supported on Services. This will have 2 potential values:
+
+- `Auto`: When there are a sufficient number of endpoints for the Service, the
+  EndpointSlice controller will add topology hints for each endpoint that will
+  ensure a proportional amounts are available to each zone in a cluster.
+- `Disabled`: EndpointSlice hints will not be set for this feature.
+
+When this annotation is either unspecified or not set to `Auto`, the `Disabled`
+behavior will be default. A future KEP will explore changing the default
+behavior from `Disabled` to `Auto` with a new feature gate.
+
+#### Interoperability
+
+If both this annotation and `topologyKeys` are set on a Service, `topologyKeys`
+will be given precedence and this annotation will be ignored.
+
+#### Feature Gate
+
+This functionality will be guarded by the `TopologyAwareHints` feature
+gate. This gate should not be enabled if the `ServiceTopology` gate is enabled.
+
+### API
+
+A new `TopologyHints` struct would be added to the `EndpointSlice.Endpoint`
+struct:
+
+```go
+type Endpoint struct {
+  ...
+  // hints contains information associated with how an endpoint should be
+  // consumed.
+  // +optional
+  Hints EndpointHints `json:"hints,omitempty" protobuf:"bytes,7,opt,name=hints"`
+}
+```
+
+```go
+// EndpointHints provides hints describing how an endpoint should be consumed.
+type EndpointHints struct {
+  // zone indicates the zone this endpoint should be consumed by to enable
+  // topology aware routing.
+  // +optional
+  Zone string `json:"ready,omitempty" protobuf:"bytes,1,name=ready"`
+}
+```
+
+#### Future API Expansion
+This approach would allow for future API expansion that enabled specifying
+multiple zones per endpoint with weights. That level of complexity may never be
+necessary, but it will be possible. For example:
+
+```yaml
+topologyHints:
+  zones:
+  - name: example-1a
+    weight: 50
+  - name: example-2a
+    weight: 50
+```
+
+Additionally we could easily expand this API to include support for region
+hints. Although it is unclear if either expansion will be necessary, the API is
+designed in a way to make expansions straightforward.
+
+### Kube-Proxy
+
+When the `TopologyAwareHints` feature gate is enabled, Kube-Proxy will be
```

Comment:
I don't remember why, but for slices I think you ended having two independent feature gates for kube-proxy and the slice controller, can this happen again?

Reply:
Yeah I almost did that here. In that case I actually split the features up into different KEPs entirely because they felt sufficiently different. This feels more like a single feature, but I'm open to a separate feature gate here. Any naming recommendations?

Reply:
I was just asking because I don't remember the cause, so if we don't know I don't think that is a big deal to amend it later if we find out it is really needed